tokio/src/task/core.rs

```diff
@@ -55,102 +54,213 @@ pub(crate) struct Header {
 
     /// Table of function pointers for executing actions on the task.
     pub(super) vtable: &'static Vtable,
-
-    /// Used by loom to track the causality of the future. Without loom, this is
-    /// unit.
-    pub(super) future_causality: CausalCell<()>,
 }
 
 /// Cold data is stored after the future.
 pub(super) struct Trailer {
     /// Consumer task waiting on completion of this task.
-    pub(super) waker: CausalCell<MaybeUninit<Option<Waker>>>,
+    pub(super) waker: CausalCell<Option<Waker>>,
 }
 
 /// Either the future or the output.
-enum Stage<T: Future> {
-    Running(Track<T>),
-    Finished(Track<super::Result<T::Output>>),
+pub(super) enum Stage<T: Future> {
+    Running(T),
+    Finished(super::Result<T::Output>),
     Consumed,
 }
 
-impl<T: Future> Cell<T> {
+impl<T: Future, S: Schedule> Cell<T, S> {
     /// Allocates a new task cell, containing the header, trailer, and core
     /// structures.
-    pub(super) fn new<S>(future: T, state: State) -> Box<Cell<T>>
-    where
-        S: Schedule,
-    {
+    pub(super) fn new(future: T, state: State) -> Box<Cell<T, S>> {
         Box::new(Cell {
             header: Header {
                 state,
-                executor: CausalCell::new(None),
-                queue_next: UnsafeCell::new(ptr::null()),
+                queue_next: UnsafeCell::new(None),
                 owned_next: UnsafeCell::new(None),
                 owned_prev: UnsafeCell::new(None),
                 vtable: raw::vtable::<T, S>(),
-                future_causality: CausalCell::new(()),
             },
             core: Core {
-                stage: Stage::Running(Track::new(future)),
+                scheduler: CausalCell::new(None),
+                stage: CausalCell::new(Stage::Running(future)),
             },
             trailer: Trailer {
-                waker: CausalCell::new(MaybeUninit::new(None)),
+                waker: CausalCell::new(None),
             },
         })
     }
 }
 
-impl<T: Future> Core<T> {
-    pub(super) fn transition_to_consumed(&mut self) {
-        self.stage = Stage::Consumed
+impl<T: Future, S: Schedule> Core<T, S> {
+    /// If needed, bind a scheduler to the task.
+    ///
+    /// This only happens on the first poll.
+    pub(super) fn bind_scheduler(&self, task: Task<S>) {
+        use std::mem::ManuallyDrop;
+
+        // TODO: it would be nice to not have to wrap with a ManuallyDrop
```

Comment:
Why is it necessary? If `S::bind` panics, the task will be dropped -- is that an issue? If it is, should this `fn` just take `&mut Task`?

Reply:
Yes, these bits aren't super clean. I opted to punt further cleanup of this until another PR.